go/libkb/client.go

```diff
@@ -191,20 +191,14 @@ func NewClient(g *GlobalContext, config *ClientConfig, needCookie bool) *Client
 		return c, nil
 	}
 
-	if (config != nil && config.RootCAs != nil) || env.GetTorMode().Enabled() {
-		if config != nil && config.RootCAs != nil {
-			xprt.TLSClientConfig = &tls.Config{RootCAs: config.RootCAs}
-		}
-		if env.GetTorMode().Enabled() {
-			extraLog(nil, "api.Client:%v tor mode enabled", needCookie)
-			// TODO: should we call res_init on DNS errors here as well?
-			dialSocksProxy := socks.DialSocksProxy(socks.SOCKS5, env.GetTorProxy())
-			xprt.DialContext = func(ctx context.Context, network, addr string) (c net.Conn, err error) {
-				return dialSocksProxy(network, addr)
-			}
-		} else {
-			xprt.Proxy = http.ProxyFromEnvironment
-		}
+	if config != nil && config.RootCAs != nil {
+		xprt.TLSClientConfig = &tls.Config{RootCAs: config.RootCAs}
+	}
+
+	err := EnableProxy(env.GetProxyType(), env.GetProxy())
```

Comment:
can we achieve the same effect by just fiddling with `xprt` above? It feels kludgy to change this local behavior via global variables. I think from the docs we can

https://golang.org/src/net/http/transport.go?s=3607:10104#L93

Reply:
Yes, it is possible to change it via setting the Proxy function. I initially did it this way because [`s3.go`](https://github.com/keybase/client/blob/master/go/chat/s3/s3.go#L953) and [`search.go`](https://github.com/keybase/client/blob/master/go/chat/giphy/search.go#L114) currently do not use the http client created in `client.go`. By setting it in the environment variables, those http.Clients (and any others that are made in the future) will automatically get configured to use the proxy. 

I could either refactor `s3.go` and `search.go` to use `client.go` as the HTTP client or I could simply modify them to use the proxy directly. What are your thoughts on the best solution here? I'm leaning towards migrating them to reference `client.go` but I'm not sure if there is a reason they aren't using it already.

Reply:
I wouldn't refactor, I would just modify the other two instances individually. There is likely some good reason they weren't using `client.go` to begin with. Cc @mmaxim  and @patrickxb  on why not.

Reply:
chat/s3/s3.go is for chat attachment uploads and uses its own http client.  We heavily modified that package for direct to s3 uploads.

I would prefer if it was modified as little as possible as it has been working well for years.

Reply:
@patrickxb  What do you think of this solution? I attempted to keep the changes to s3 as minimal as possible but it required some decent changes to wind the global context in in order to get access to the config.